akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -185,3 +191,143 @@ private[akka] final class SourceQueueAdapter[T](delegate: SourceQueueWithComplet
   def complete(): Unit = delegate.complete()
   def fail(ex: Throwable): Unit = delegate.fail(ex)
 }
+
+/**
+ * INTERNAL API
+ */
+private[stream] final class UnfoldResourceSource[T, S](create: () ⇒ S,
+                                                       readData: (S) ⇒ Option[T],
+                                                       close: (S) ⇒ Unit) extends GraphStage[SourceShape[T]] {
+  val out = Outlet[T]("UnfoldResourceSource.out")
+  override val shape = SourceShape(out)
+  override def initialAttributes: Attributes = DefaultAttributes.unfoldResourceSource
+
+  def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with OutHandler {
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    var blockingStream: S = _
+    setHandler(out, this)
+
+    override def preStart(): Unit = blockingStream = create()
+
+    @tailrec
+    final override def onPull(): Unit = {
+      var resumingMode = false
+      try {
+        readData(blockingStream) match {
+          case Some(data) ⇒ push(out, data)
+          case None       ⇒ closeStage()
+        }
+      } catch {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒
+            close(blockingStream)
+            failStage(ex)
+          case Supervision.Restart ⇒
+            restartState()
+            resumingMode = true
+          case Supervision.Resume ⇒
+            resumingMode = true
+        }
+      }
+      if (resumingMode) onPull()
+    }
+
+    override def onDownstreamFinish(): Unit = closeStage()
```

Comment:
My assumption on behavior is: In case close or open throw exception - stream should failed. 
But with having close called in postStop stream does not failed it completes nomally.
Let me know if we should rethink this.